fuse_core/src/async_publisher.cpp

```diff
@@ -33,71 +33,77 @@
-#include <fuse_core/graph.h>
-#include <fuse_core/transaction.h>
-#include <ros/node_handle.h>
-
-#include <boost/make_shared.hpp>
-
-#include <functional>
-#include <utility>
-#include <string>
+#include <rclcpp/contexts/default_context.hpp>
 
 
 namespace fuse_core
 {
 
 AsyncPublisher::AsyncPublisher(size_t thread_count) :
   name_("uninitialized"),
-  spinner_(thread_count, &callback_queue_)
+  executor_thread_count_(thread_count)
 {
 }
 
 void AsyncPublisher::initialize(const std::string& name)
 {
   // Initialize internal state
   name_ = name;
-  node_handle_.setCallbackQueue(&callback_queue_);
-  private_node_handle_ = ros::NodeHandle("~/" + name_);
-  private_node_handle_.setCallbackQueue(&callback_queue_);
+  std::string node_namespace = "";
+
+  // TODO(CH3): Pass in the context or a node to get the context from
+  rclcpp::Context::SharedPtr ros_context = rclcpp::contexts::get_global_default_context();
+  auto node_options = rclcpp::NodeOptions();
+  node_options.context(ros_context); //set a context to generate the node in
+
+  // TODO(CH3): Potentially pass in the optimizer node instead of spinning a new one
+  node_ = rclcpp::Node::make_shared(name_, node_namespace, node_options);
+
+  auto executor_options = rclcpp::ExecutorOptions();
+  executor_options.context = ros_context;
+  executor_ = rclcpp::executors::MultiThreadedExecutor::make_shared(executor_options, executor_thread_count_);
+
+  callback_queue_ = std::make_shared<CallbackAdapter>(ros_context);
+  node_->get_node_waitables_interface()->add_waitable(
+    callback_queue_, (rclcpp::CallbackGroup::SharedPtr) nullptr);
 
   // Call the derived onInit() function to perform implementation-specific initialization
   onInit();
 
-  // Start the async spinner to service the local callback queue
-  spinner_.start();
+  executor_->add_node(node_);
 }
 
 void AsyncPublisher::notify(Transaction::ConstSharedPtr transaction, Graph::ConstSharedPtr graph)
 {
   // Insert a call to the `notifyCallback` method into the internal callback queue.
   // This minimizes the time spent by the optimizer's thread calling this function.
-  auto callback = boost::make_shared<CallbackWrapper<void>>(
+  auto callback = std::make_shared<CallbackWrapper<void>>(
     std::bind(&AsyncPublisher::notifyCallback, this, std::move(transaction), std::move(graph)));
-  callback_queue_.addCallback(callback, reinterpret_cast<uint64_t>(this));
+  callback_queue_->addCallback(callback);
 }
 
 void AsyncPublisher::start()
 {
-  auto callback = boost::make_shared<CallbackWrapper<void>>(std::bind(&AsyncPublisher::onStart, this));
-  auto result = callback->getFuture();
-  callback_queue_.addCallback(callback, reinterpret_cast<uint64_t>(this));
-  result.wait();
+  auto callback = std::make_shared<CallbackWrapper<void>>(std::bind(&AsyncPublisher::onStart, this));
+  // auto result = callback->getFuture();  // TODO(CH3): Circle back if we need this again
+  callback_queue_->addCallback(callback);
+  // result.wait();
```

Comment:
Same comment about `start/stop` being synchronous in the ROS 1 version

Reply:
https://github.com/locusrobotics/fuse/compare/d3b6c944272008f0ad4f62582c73deb350f3b9e1..3e38f8d4c103da15b673464842e504c4d7a8f799